File: trimet/stops.js
'use strict';

const fs = require('fs');
const path = require('path');
const axios = require('axios');
const api = require('./api');

module.exports = {
  getStops,
};

const DATA_DIR = path.join(__dirname, 'data');

/**
 * Fetch stops and routes from TriMet API and save as JSON.
 *
 * This fetches all stops and their associated routes. Stops and routes
 * are saved to separate JSON files in the specified directory.
 *
 * @param directory {String} Directory to save stops & routes data into
 * @param quiet {Boolean} Squelch info messages
 * @returns {Promise}
 */
function getStops(directory = DATA_DIR, quiet = false) {
  const [url, options] = api.getRequestOptions('stops', {
    ll: '-122.667369,45.522698',
    feet: 5280 * 100,
    showRoutes: true,
    showRouteDirs: true,
  });
  const stopsFile = path.join(directory, 'stops.json');
  const rawStopsFile = path.join(directory, 'rawStops.json');
  const routesFile = path.join(directory, 'routes.json');

  function info(...args) {
    if (!quiet) {
      console.info(...args);
    }
  }

  function process(data) {
    const root = data.resultSet;
    const results = root.location;
    const queryTime = root.queryTime;
    const stops = [];
    const routes = [];
    const seenRoutes = new Set();

    function stringify(items) {
      items.sort((a, b) => a.id - b.id);
      return JSON.stringify(
        { retrieved: queryTime, data: items },
        undefined,
        2
      );
    }

    info(`Processing ${results.length} stops...`);

    for (let result of results) {
      const stopID = result.locid;
      const seenStopRoutes = new Set();
      const stopRoutes = [];
      const stop = {
        id: stopID,
        name: result.desc,
        direction: result.dir || 'unknown',
        location: [result.lng, result.lat],
        routes: stopRoutes,
      };

      stops.push(stop);

      for (let routeResult of result.route || []) {
        const routeID = routeResult.route;

        for (let dir of routeResult.dir) {
          const direction = dir.dir === 0 ? 'outbound' : 'inbound';
          const description = dir.desc;
          const routeKey = `${routeID}:${direction}`;
          const stopRouteKey = `${stopID}:${routeID}:${direction}`;

          if (!seenRoutes.has(routeKey)) {
            const type = routeResult.type;
            const name = routeResult.desc;
            const route = { id: routeID, direction, type, name, description };
            route.type = getType(route);
            route.short_name = getShortName(route);
            seenRoutes.add(routeKey);
            routes.push(route);
          }

          if (!seenStopRoutes.has(stopRouteKey)) {
            seenStopRoutes.add(stopRouteKey);
            stopRoutes.push({ id: routeID, direction });
          }
        }
      }
    }

    info(`Writing stops...`);
    fs.writeFileSync(stopsFile, stringify(stops));
    info(`Stops saved to ${stopsFile}`);

    info(`Writing routes...`);
    fs.writeFileSync(routesFile, stringify(routes));
    info(`Routes saved to ${routesFile}`);

    return [stops, routes];
  }

  let data;
  try {
    data = fs.readFileSync(rawStopsFile);
  } catch (error) {
    if (error.code !== 'ENOENT') {
      // If the file doesn't exist, that's okay, but any other error
      // should be re-thrown.
      throw error;
    }
  }

  if (data) {
    info(`Got stops from local file ${rawStopsFile}`);
    data = JSON.parse(data);
    data = process(data);
    return Promise.resolve(data);
  }

  return axios
    .get(url, options)
    .then(response => {
      let { data } = response;
      if (data.error) {
        throw new Error(data.error.content || 'Unknown error');
      }
      info(`Got stops from TriMet API; saving raw data to ${rawStopsFile}`);
      fs.writeFileSync(rawStopsFile, JSON.stringify(data, undefined, 2));
      return process(data);
    })
    .catch(err => {
      console.error(`Could not get stops: ${err.message}`);
      throw err;
    });
}

function getType(route) {
  const { id, type, name } = route;
  if (type === 'B') {
    return name.endsWith('Shuttle') ? 'shuttle' : 'bus';
  } else if (type === 'R') {
    if (name.startsWith('MAX')) {
      return 'light-rail';
    } else if (name.startsWith('Portland Streetcar')) {
      return 'streetcar';
    } else if (name.startsWith('Aerial Tram')) {
      return 'aerial-tram';
    } else if (name.startsWith('WES')) {
      return 'commuter-rail';
    }
    throw new Error(`Could not determine fixed route type for ${id}: ${type}`);
  }
  throw new Error(`Unexpected route type for ${id}: ${type}`);
}

function getShortName(route) {
  const { id, name, type } = route;
  switch (type) {
    case 'bus':
      return id.toString();
    case 'shuttle':
      return name.split(' - ')[1] || name;
    case 'light-rail':
      return name.replace(/^MAX\s+/, '').replace(/\s+Line$/, '');
    case 'streetcar':
      return name.split(' - ')[1];
    case 'aerial-tram':
      return 'Aerial Tram';
    case 'commuter-rail':
      return 'WES';
    default:
      throw new Error(`Could not determine short name for ${id}: ${type}`);
  }
}

File: src/api/mystops/trimet/stops.py
import json
import re
from pathlib import Path

from .request import make_request


def get_stops(api_key, out_dir, overwrite=False):
    """Fetch stops and routes from TriMet API and save as JSON.

    This fetches all stops and their associated routes. Stops and routes
    are saved to separate JSON files in the specified directory.

    The raw data from the TriMet API is also saved into the output
    directory. If the raw data file exists

    """
    out_dir = Path(out_dir)

    # Data from TriMet API as-is
    raw_stops_file = out_dir / "raw_stops.json"

    stops_file = out_dir / "stops.json"
    routes_file = out_dir / "routes.json"

    if not out_dir.exists():
        print(f"Creating output directory: {out_dir}")
        out_dir.mkdir(parents=True)

    if raw_stops_file.exists() and not overwrite:
        print("Reading cached stop data from disk")
        with raw_stops_file.open() as fp:
            data = fp.read()
            data = json.loads(data)
    else:
        print("Fetching stop data from TriMet API")
        response = make_request(
            "stops",
            api_key,
            params={
                "ll": "-122.667369,45.522698",
                "feet": 5280 * 100,
                "showRoutes": "true",
                "showRouteDirs": "true",
            },
        )
        with raw_stops_file.open("w") as fp:
            fp.write(response.text)
        data = response.json()

    root = data["resultSet"]
    results = root["location"]
    query_time = root["queryTime"]
    stops = []
    routes = []
    seen_routes = set()

    for result in results:
        stop_id = result["locid"]
        seen_stop_routes = set()
        stop_routes = []

        stop = {
            "id": stop_id,
            "name": result["desc"],
            "direction": result.get("dir") or None,
            "location": (result["lng"], result["lat"]),
            "routes": stop_routes,
        }

        stops.append(stop)

        route = result.get("route") or ()
        for route_result in route:
            route_id = route_result["route"]
            for dir in route_result["dir"]:
                dir_code = dir["dir"]
                direction = "outbound" if dir_code == 0 else "inbound"
                description = dir["desc"]
                route_key = (route_id, direction)
                stop_route_key = (stop_id, route_id, direction)
                if route_key not in seen_routes:
                    type = route_result["type"]
                    name = route_result["desc"]
                    route = {
                        "id": route_id,
                        "direction": direction,
                        "type": type,
                        "name": name,
                        "description": description,
                    }
                    route["type"] = get_route_type(route)
                    route["short_name"] = get_route_short_name(route)
                    seen_routes.add(route_key)
                    routes.append(route)
                if stop_route_key not in seen_stop_routes:
                    seen_stop_routes.add(stop_route_key)
                    stop_routes.append({"id": route_id, "direction": direction})

    print("Writing stops")
    with stops_file.open("w") as fp:
        stops.sort(key=lambda s: s["id"])
        data = json.dumps(
            {
                "retrieved": query_time,
                "data": stops,
            }
        )
        fp.write(data)

    print("Writing routes")
    with routes_file.open("w") as fp:
        routes.sort(key=lambda r: r["id"])
        data = json.dumps(
            {
                "retrieved": query_time,
                "data": routes,
            }
        )
        fp.write(data)


def get_route_type(route):
    id = route["id"]
    type = route["type"]
    name = route["name"]
    if type == "B":
        if name.endswith("Shuttle"):
            return "shuttle"
        else:
            return "bus"
    if type == "R":
        if name.startswith("MAX"):
            return "light-rail"
        elif name.startswith("Portland Streetcar"):
            return "streetcar"
        elif name.startswith("Aerial Tram"):
            return "aerial-tram"
        elif name.startswith("WES"):
            return "commuter-rail"
        else:
            raise ValueError(f"Could not determine fixed route type for {id}: {type}")
    raise ValueError(f"Unexpected route type for {id}: {type}")


def get_route_short_name(route):
    id = route["id"]
    type = route["type"]
    name = route["name"]
    if type == "bus":
        return str(id)
    elif type == "shuttle":
        parts = name.split(" - ")
        if len(parts) > 1:
            return parts[1]
        return name
    elif type == "light-rail":
        value = re.sub(r"^MAX\s+", "", name)
        value = re.sub(r"\s+Line$", "", name)
        return value
    elif type == "streetcar":
        return name.split(" - ")[1]
    elif type == "aerial-tram":
        return "Aerial Tram"
    elif type == "commuter-rail":
        return "WES"
    raise ValueError(f"Could not determine short name for {id}: {type}")
